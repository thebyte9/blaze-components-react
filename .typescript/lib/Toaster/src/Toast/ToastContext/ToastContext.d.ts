declare const ToastContext: any;
export default ToastContext;
