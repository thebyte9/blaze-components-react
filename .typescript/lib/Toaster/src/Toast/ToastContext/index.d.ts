import ToastContext from "./ToastContext";
export default ToastContext;
