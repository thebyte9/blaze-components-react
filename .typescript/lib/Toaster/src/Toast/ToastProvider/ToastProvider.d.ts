import { Component } from "react";
import { Callback, IOptions } from "../types/common";
import { IProps, IState } from "../types/provider/provider";
declare class ToastProvider extends Component<IProps, IState> {
    static defaultProps: {
        autoDismiss: boolean;
        autoDismissTimeout: number;
        components: {
            Toast: {
                ({ appearance, autoDismiss, autoDismissTimeout, children, isRunning, onDismiss, placement, transitionDuration, transitionState, onMouseEnter, onMouseLeave, ...otherProps }: import("../types/common").IToastProps): JSX.Element;
                defaultProps: {
                    onDismiss: () => void;
                };
            };
            ToastContainer: ({ hasToasts, placement, ...props }: any) => JSX.Element;
        };
        placement: string;
        transitionDuration: number;
    };
    state: {
        toasts: never[];
    };
    has: (id: string) => boolean;
    onDismiss: (id: string, cb?: Callback) => () => void;
    add: (content: JSX.Element, options?: IOptions, cb?: Callback) => any;
    remove: (id: string, cb?: Callback) => void;
    removeAll: () => void;
    update: (id: string, options?: IOptions, cb?: Callback) => void;
    render(): JSX.Element;
}
export default ToastProvider;
