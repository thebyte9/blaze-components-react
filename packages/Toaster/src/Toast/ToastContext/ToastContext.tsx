import React from "react";

const ToastContext: any = React.createContext({});

export default ToastContext;
