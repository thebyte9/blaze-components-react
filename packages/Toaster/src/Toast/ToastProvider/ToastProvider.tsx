import React, { Component } from "react";
import { createPortal } from "react-dom";
import { Transition, TransitionGroup } from "react-transition-group";
import {
  DEFAULT_AUTO_DISMISS_TIMEOUT,
  DEFAULT_PLACEMENT,
  DEFAULT_TRANSITION_DURATION
} from "../constants";
import DefaultToast from "../DefaultToast";
import ToastContainer from "../ToastContainer/ToastContainer";
import ToastContext from "../ToastContext";
import ToastController from "../ToastController";
import { Callback, Id, IOptions } from "../types/common";
import { IProps, IState } from "../types/provider/provider";
import { generateUEID, NOOP } from "../utils";

const canUseDOM = !!(
  typeof window !== "undefined" &&
  window.document &&
  window.document.createElement
);

class ToastProvider extends Component<IProps, IState> {
  public static defaultProps = {
    autoDismiss: false,
    autoDismissTimeout: DEFAULT_AUTO_DISMISS_TIMEOUT,
    components: { Toast: DefaultToast, ToastContainer },
    placement: DEFAULT_PLACEMENT,
    transitionDuration: DEFAULT_TRANSITION_DURATION
  };

  public state = { toasts: [] };

  public has = (id: string) => {
    if (!this.state.toasts.length) {
      return false;
    }

    return Boolean(this.state.toasts.filter((t: any) => t.id === id).length);
  };
  public onDismiss = (id: Id, cb: Callback = NOOP) => () => {
    cb(id);
    this.remove(id);
  };

  public add = (
    content: JSX.Element,
    options: IOptions = {},
    cb: Callback = NOOP
  ) => {
    const id = options.id || generateUEID();
    const callback = () => cb(id);

    if (this.has(id)) {
      return;
    }

    this.setState((state: any): any => {
      const newToast = { content, id, ...options };
      const toasts = [...state.toasts, newToast];

      return { toasts };
    }, callback);

    return id;
  };
  public remove = (id: Id, cb: Callback = NOOP) => {
    const callback = () => cb(id);

    if (!this.has(id)) {
      return;
    }

    this.setState(state => {
      const toasts = state.toasts.filter(t => t.id !== id);
      return { toasts };
    }, callback);
  };
  public removeAll = () => {
    if (!this.state.toasts.length) {
      return;
    }

    this.state.toasts.forEach((t: any) => this.remove(t.id));
  };
  public update = (id: Id, options: IOptions = {}, cb: Callback = NOOP) => {
    const callback = () => cb(id);

    if (!this.has(id)) {
      return;
    }

    this.setState(state => {
      const old = state.toasts;
      const i = old.findIndex(t => t.id === id);
      const updatedToast = { ...old[i], ...options };
      const toasts = [...old.slice(0, i), updatedToast, ...old.slice(i + 1)];

      return { toasts };
    }, callback);
  };

  public render() {
    const {
      autoDismiss: inheritedAutoDismiss,
      autoDismissTimeout,
      children,
      components,
      placement,
      transitionDuration
    } = this.props;
    const { Toast, ToastContainer } = { ...components };
    const { add, remove, removeAll, update } = this;
    const toasts = Object.freeze(this.state.toasts);

    const hasToasts = Boolean(toasts.length);
    const portalTarget = canUseDOM ? document.body : null;

    return (
      <ToastContext.Provider value={{ add, remove, removeAll, update, toasts }}>
        {children}

        {portalTarget ? (
          createPortal(
            <ToastContainer placement={placement} hasToasts={hasToasts}>
              <TransitionGroup component={null}>
                {toasts.map(
                  ({
                    appearance,
                    autoDismiss,
                    content,
                    id,
                    onDismiss,
                    ...unknownConsumerProps
                  }: any) => (
                    <Transition
                      appear
                      key={id}
                      mountOnEnter
                      timeout={transitionDuration}
                      unmountOnExit
                    >
                      {transitionState => (
                        <ToastController
                          appearance={appearance}
                          autoDismiss={
                            autoDismiss !== undefined
                              ? autoDismiss
                              : inheritedAutoDismiss
                          }
                          autoDismissTimeout={autoDismissTimeout}
                          component={Toast}
                          key={id}
                          onDismiss={this.onDismiss(id, onDismiss)}
                          placement={placement}
                          transitionDuration={transitionDuration}
                          transitionState={transitionState}
                          {...unknownConsumerProps}
                        >
                          {content}
                        </ToastController>
                      )}
                    </Transition>
                  )
                )}
              </TransitionGroup>
            </ToastContainer>,
            portalTarget
          )
        ) : (
          <ToastContainer placement={placement} hasToasts={hasToasts} /> // keep ReactDOM.hydrate happy
        )}
      </ToastContext.Provider>
    );
  }
}

export default ToastProvider;
