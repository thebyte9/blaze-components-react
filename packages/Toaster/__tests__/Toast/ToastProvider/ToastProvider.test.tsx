import { act, cleanup, render } from '@testing-library/react';

import React from 'react';
import ToastContext from '../../../src/Toast/ToastContext';
import ToastProvider from '../../../src/Toast/ToastProvider/ToastProvider';
import expect from 'expect';

jest.mock('../../../src/Toast/ToastContext', () => ({
  Provider: jest.fn(({ children }) => children),
}));

describe('Toast provider', () => {
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  it('should be defined', () => {
    expect(ToastProvider).toBeDefined();
  });

  it('should render without throwing error', () => {
    const { container } = render(
      <ToastProvider>
        <div>children toastprovider component</div>
      </ToastProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should add a new toaster', () => {
    const FakeToasterTitleImplementation = 'Test toaster';
    const FakeToasterConfigImplementation = {
      appearance: 'success',
      autoDismiss: true,
    };
    const FakeChildrenComponentImplementation: any = jest.fn(({ children: subChildren }) => <div>{subChildren}</div>);

    render(
      <ToastProvider>
        <FakeChildrenComponentImplementation />
      </ToastProvider>,
    );

    const [
      [
        {
          value: { add },
        },
      ],
    ] = ToastContext.Provider.mock.calls;

    act(() => {
      add(FakeToasterTitleImplementation, FakeToasterConfigImplementation);
    });

    const [
      ,
      [
        {
          value: { toasts },
          children,
        },
      ],
    ] = ToastContext.Provider.mock.calls;

    expect(toasts).toStrictEqual([
      expect.objectContaining({
        content: FakeToasterTitleImplementation,
        ...FakeToasterConfigImplementation,
      }),
    ]);
    expect(children).toMatchSnapshot();
  });
});
