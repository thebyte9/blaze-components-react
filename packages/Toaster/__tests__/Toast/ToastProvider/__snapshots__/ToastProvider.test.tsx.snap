// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast provider should add a new toaster 1`] = `
Array [
  <mockConstructor />,
  Object {
    "$$typeof": Symbol(react.portal),
    "children": <ToastContainer
      hasToasts={true}
      placement="top-right"
    >
      <TransitionGroup
        childFactory={[Function]}
        component={null}
      >
        <Transition
          appear={true}
          enter={true}
          exit={true}
          in={false}
          mountOnEnter={true}
          onEnter={[Function]}
          onEntered={[Function]}
          onEntering={[Function]}
          onExit={[Function]}
          onExited={[Function]}
          onExiting={[Function]}
          timeout={220}
          unmountOnExit={true}
        >
          [Function]
        </Transition>
      </TransitionGroup>
    </ToastContainer>,
    "containerInfo": <body>
      <div>
        <div />
      </div>
      <div
        class="react-toast-notifications__container"
        style="top: 0px; right: 0px;"
      >
        <div
          style="height: auto; transition: height 120ms 100ms;"
        >
          <div
            class="react-toast-notifications__toast react-toast-notifications__toast--success"
            style="transition: transform 220ms cubic-bezier(0.2, 0, 0, 1), opacity 220ms; transform: translate3d(120%, 0, 0);"
          >
            <div
              class="react-toast-notifications__toast__icon-wrapper react-toast-notifications__toast__icon-wrapper--success"
            >
              <div
                class="react-toast-notifications__toast__countdown_animation"
                style="animation-duration: 5000ms; animation-play-state: running; opacity: 1;"
              />
              <svg
                aria-hidden="true"
                class="react-toast-notifications__toast__icon"
                fill="#FFFFFF"
                height="16"
                style="display: inline-block; vertical-align: text-top;"
                viewBox="0 0 12 16"
                width="12"
              >
                <path
                  d="M12 5.5l-8 8-4-4L1.5 8 4 10.5 10.5 4 12 5.5z"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
            <div
              class="react-toast-notifications__toast__content"
            >
              Test toaster
            </div>
            <div
              class="react-toast-notifications__toast__dismiss-button"
              role="button"
            >
              <svg
                aria-hidden="true"
                class="react-toast-notifications__toast__dismiss-icon react-toast-notifications__toast__dismiss-icon--success"
                fill="currentColor"
                height="16"
                style="display: inline-block; vertical-align: text-top;"
                viewBox="0 0 14 16"
                width="14"
              >
                <path
                  d="M7.71 8.23l3.75 3.75-1.48 1.48-3.75-3.75-3.75 3.75L1 11.98l3.75-3.75L1 4.48 2.48 3l3.75 3.75L9.98 3l1.48 1.48-3.75 3.75z"
                  fill-rule="evenodd"
                />
              </svg>
              <span
                class="react-toast-notifications__toast__dismiss-text"
              />
            </div>
          </div>
        </div>
      </div>
    </body>,
    "implementation": null,
    "key": null,
  },
]
`;

exports[`Toast provider should render without throwing error 1`] = `<div />`;
